simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs

```diff
@@ -1,275 +1,409 @@
+        /// </summary>
+        /// <returns>data after a possible commit</returns>
+        public ObservableCollection<T> GetAsObservableCollection()
+        {
+            ObservableCollection<T> preview = new ObservableCollection<T>(this.items);
+            foreach (var item in this.newItems)
+            {
+                preview.Add(item);
+            }
+            return preview;
+        }
+
+        /// <summary>
+        /// Add a new item.
+        /// If the item already exists - ignore it
+        /// If the item was removed, put it in the item list again
+        /// </summary>
+        /// <param name="item">Item to add</param>
+        public void Add(T item)
+        {
+            // If the item is not added to any list yet
+            if (removedItems.Contains(item))
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item,
+                    ReaddFromRemovedItems = true
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
+                {
+                    newItems.Add(item);
+                    // Remove item from the list of removed items
+                    items.Add(item);
+                    removedItems.Remove(item);
+
+                    var itemAdded = new ItemAddedEventArgs
+                    {
+                        Item = item
+                    };
+                    ItemAdded?.Invoke(this, itemAdded);
+                }
+            }
+            else if (!items.Contains(item) && !newItems.Contains(item))
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
+                {
+                    newItems.Add(item);
+
+                    var itemAdded = new ItemAddedEventArgs
+                    {
+                        Item = item
+                    };
+                    ItemAdded?.Invoke(this, itemAdded);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add a range of items. Each item will be handled as described for <see cref="Add(T)"/>
+        /// </summary>
+        /// <param name="items">List of items to handle</param>
+        public void AddRange(IEnumerable<T> collection)
+        {
+            foreach (var item in collection)
+            {
+                Add(item);
+            }
+        }
+
+        /// <summary>
+        /// Clear the howle collection
+        /// </summary>
+        public void Clear()
+        {
+            var args = new ClearCollectionEventArgs();
+            ClearCollection?.Invoke(this, args);
+
+            if (!args.Cancel)
+            {
+                items.Clear();
+                newItems.Clear();
+                removedItems.Clear();
+
+                var collectionClearedArgs = new CollectionClearedEventArgs();
```

Comment:
maybe put counters in the args: ItemsCleared: 45, NewItemsCleared: 2, RemovedItemsCleared: 3